clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+     * regardless of the set of "lost partitions."
+     *
+     * @param lostPartitions Set of {@link TopicPartition partitions} that were lost
+     * @see AssignmentReconciler
+     */
+    void completeLost(Set<TopicPartition> lostPartitions) {
+        log.debug("{} callbacks were successfully invoked", ConsumerRebalanceListener.class.getSimpleName());
+        subscriptions.assignFromSubscribed(Collections.emptySet());
+    }
+
+    /**
+     * Performs the assignment phase of the reconciliation process as described in the top-level class documentation.
+     *
+     * @param assignment Holds the {@link Assignment} which includes the target set of topics
+     */
+
+    void startReconcile(Assignment assignment) {
+        SortedSet<TopicPartition> partitionsToRevoke = getPartitionsToRevoke(assignment);
+        SortedSet<TopicPartition> partitionsToAssign = getPartitionsToAssign(assignment);
+
+        if (partitionsToRevoke.isEmpty() && partitionsToAssign.isEmpty()) {
+            log.debug("Skipping invocation of {} callbacks as no partitions changed in the new assignment",
+                    ConsumerRebalanceListener.class.getSimpleName());
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        PartitionReconciliationStartedEvent event = new PartitionReconciliationStartedEvent(
+                partitionsToRevoke,
+                partitionsToAssign
+        );
+        backgroundEventQueue.add(event);
+    }
+
+    /**
+     * This method should be invoked to signal the completion of a successful {@link TopicPartition partition}
+     * assignment reconciliation. Specifically, it is to be executed on background thread <em>after</em> the
+     * {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} and
+     * {@link ConsumerRebalanceListener#onPartitionsAssigned(Collection)} callbacks have completed execution on
+     * the application thread. It should update the set of
+     * {@link SubscriptionState#assignedPartitions() assigned partitions} based on the
+     * given partitions.
+     *
+     * @param revokedPartitions Set of {@link TopicPartition partitions} that were revoked
+     * @param assignedPartitions Set of {@link TopicPartition partitions} that were assigned
+     * @see MembershipManager#completeReconcile(Set, Set, Optional)
+     */
+    void completeReconcile(Set<TopicPartition> revokedPartitions, Set<TopicPartition> assignedPartitions) {
+        log.debug("{} callbacks were successfully invoked", ConsumerRebalanceListener.class.getSimpleName());
+
+        Set<TopicPartition> newAssignment = new HashSet<>(subscriptions.assignedPartitions());
+        newAssignment.addAll(assignedPartitions);
+        newAssignment.removeAll(revokedPartitions);
+        subscriptions.assignFromSubscribed(newAssignment);
```

Comment:
For my understanding, I suppose that the fetcher will start to fetch from the newly assigned partitions after this point. Is my understanding right?

Reply:
Yes, though not immediately. When the fetcher does its next pass and sees the new partitions, they will be included in the next set of outgoing fetch requests.